File: Assets/src/model/floors/WeightedRandomBag.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

/// https://gamedev.stackexchange.com/a/162996
[Serializable]
public class WeightedRandomBag<T> : IEnumerable<KeyValuePair<float, T>>, ICloneable {

  [Serializable]
  protected class Entry {
    public float weight;
    public T item;
    public override string ToString() {
      var i = item is Delegate d ? d.Method.Name : item.ToString();
      return $"{i}: {weight}";
    }
  }

  protected List<Entry> entries = new List<Entry>();

  public void Add(float weight, T item) {
    entries.Add(new Entry { item = item, weight = weight });
  }

  public void Remove(T entry) {
    entries.RemoveAll((Entry e) => e.item.Equals(entry));
  }

  public void SetWeight(T item, float weight) {
    var entry = entries.Find(e => e.item.Equals(item));
    entry.weight = weight;
  }

  public float? GetWeight(T item) {
    var entry = entries.Find(e => e.item.Equals(item));
    if (entry == null) {
      return null;
    }
    return entry.weight;
  }

  public T GetRandom() {
    var accumulatedWeight = GetAccumulatedWeight();
    /// TODO audit locations of WeightedRandomBag and make sure the stream isn't being corrupted
    float r = MyRandom.value * accumulatedWeight;
    var accum = 0f;

    foreach (Entry entry in entries) {
      accum += entry.weight;
      if (accum >= r) {
        return entry.item;
      }
    }
    return default(T); //should only happen when there are no entries
  }

  private float GetAccumulatedWeight() => entries.Select(e => e.weight).Sum();

  /// Get an item and then mutate this bag such that the item has v% less weight
  public T GetRandomAndDiscount(float reduceChanceBy = 0.3f) {
    var item = GetRandom();
    if (item != null) {
      Discount(item);
    }
    return item;
  }

  public T GetRandomAndRemove() {
    return GetRandomAndDiscount(1);
  }

  public void Discount(T item, float reduceChanceBy = 0.3f) {
    var getWeight = GetWeight(item);
    if (!getWeight.HasValue) {
      return;
    }
    var currentWeight = getWeight.Value;
    var accumulatedWeight = GetAccumulatedWeight();

    var currentChance = currentWeight / GetAccumulatedWeight();
    var newChance = currentChance * (1f - reduceChanceBy);

    // newChance = newWeight / (newWeight + rest), solve for newWeight
    // newChance * (newWeight + rest) = newWeight
    // newChance * newWeight + newChance * rest = newWeight
    // newChance * rest = newWeight - newChance * newWeight
    // newChance * rest = newWeight * (1 - newChance)
    // newChance * rest / (1 - newChance) = newWeight

    var restWeight = accumulatedWeight - currentWeight;
    var newWeight = newChance * restWeight / (1f - newChance);

    // avoid degenerate case if there's only 1 item in the bag and it's trying to discount itself
    if (newWeight == 0 && entries.Count() == 1) {
      newWeight = currentWeight;
    }
    SetWeight(item, newWeight);
  }

  public T GetRandomWithout(IEnumerable<T> encounters) {
    T pick;
    do {
      pick = GetRandom();
      if (pick == null) {
        return pick;
      }
    } while (encounters.Contains(pick));
    return pick;
  }

  public T GetRandomWithoutAndDiscount(IEnumerable<T> encounters, float reduceChanceBy = 0.3f) {
    var pick = GetRandomWithout(encounters);
    Discount(pick, reduceChanceBy);
    return pick;
  }


  public void Clear() {
    entries.Clear();
  }

  public WeightedRandomBag<T> Clone() {
    var newBag = new WeightedRandomBag<T>();
    foreach (Entry e in entries) {
      newBag.Add(e.weight, e.item);
    }
    return newBag;
  }

  public WeightedRandomBag<T> Merge(WeightedRandomBag<T> other) {
    foreach (Entry entry in other.entries) {
      var thisEntry = entries.Find(e => e.item.Equals(entry.item));
      if (thisEntry != null) {
        thisEntry.weight += entry.weight;
      } else {
        entries.Add(entry);
      }
    }
    return this;
  }

  object ICloneable.Clone() => this.Clone();

  public IEnumerator<KeyValuePair<float, T>> GetEnumerator() {
    foreach (var entry in entries) {
      yield return new KeyValuePair<float, T>(entry.weight, entry.item);
    }
  }

  IEnumerator IEnumerable.GetEnumerator() {
    return this.GetEnumerator();
  }

  public override string ToString() => $"[WeightedRandomBag {string.Join(", ", entries)}]";
}

File: web/src/generator/WeightedRandomBag.ts
import { MyRandom } from '../core/MyRandom';

/**
 * Weighted random selection with discount/subtract mechanics.
 * Port of C# WeightedRandomBag<T>.
 */
export class WeightedRandomBag<T> {
  protected entries: { weight: number; item: T }[] = [];

  add(weight: number, item: T): void {
    this.entries.push({ weight, item });
  }

  remove(item: T): void {
    this.entries = this.entries.filter(e => e.item !== item);
  }

  setWeight(item: T, weight: number): void {
    const entry = this.entries.find(e => e.item === item);
    if (entry) entry.weight = weight;
  }

  getWeight(item: T): number | null {
    const entry = this.entries.find(e => e.item === item);
    return entry ? entry.weight : null;
  }

  private getAccumulatedWeight(): number {
    return this.entries.reduce((sum, e) => sum + e.weight, 0);
  }

  getRandom(): T {
    const accumulatedWeight = this.getAccumulatedWeight();
    const r = MyRandom.value * accumulatedWeight;
    let accum = 0;
    for (const entry of this.entries) {
      accum += entry.weight;
      if (accum >= r) return entry.item;
    }
    return this.entries[this.entries.length - 1]?.item;
  }

  /** Get an item and reduce its weight by reduceChanceBy fraction */
  getRandomAndDiscount(reduceChanceBy = 0.3): T {
    // 11th hour hack - prevent drawing the same item twice
    reduceChanceBy = 1;
    const item = this.getRandom();
    if (item != null) this.discount(item, reduceChanceBy);
    return item;
  }

  getRandomAndSubtractWeight(subtract: number): T {
    const item = this.getRandom();
    if (item != null) this.subtractWeight(item, subtract);
    return item;
  }

  getRandomAndRemove(): T {
    return this.getRandomAndDiscount(1);
  }

  /**
   * Reduce item's selection probability by reduceChanceBy fraction,
   * preserving relative weights of other items.
   */
  discount(item: T, reduceChanceBy = 0.3): void {
    const currentWeight = this.getWeight(item);
    if (currentWeight == null) return;

    const accumulatedWeight = this.getAccumulatedWeight();
    const currentChance = currentWeight / accumulatedWeight;
    const newChance = currentChance * (1 - reduceChanceBy);

    // newChance = newWeight / (newWeight + rest), solve for newWeight:
    // newWeight = newChance * rest / (1 - newChance)
    const restWeight = accumulatedWeight - currentWeight;
    const newWeight = newChance * restWeight / (1 - newChance);
    if (newWeight <= 0) {
      this.remove(item);
    } else {
      this.setWeight(item, newWeight);
    }
  }

  subtractWeight(item: T, subtractWeightBy: number): void {
    const currentWeight = this.getWeight(item);
    if (currentWeight == null) return;
    if (this.entries.length === 1) return;
    this.setWeight(item, Math.max(currentWeight - subtractWeightBy, 0));
  }

  getRandomWithout(excludes: Set<T> | T[]): T {
    const excludeSet = excludes instanceof Set ? excludes : new Set(excludes);
    let pick: T;
    let guard = 0;
    do {
      pick = this.getRandom();
      if (pick == null) return pick;
    } while (excludeSet.has(pick) && ++guard < 100);
    return pick;
  }

  getRandomWithoutAndDiscount(excludes: Set<T> | T[], reduceChanceBy = 0.3): T {
    const pick = this.getRandomWithout(excludes);
    if (pick != null) this.discount(pick, reduceChanceBy);
    return pick;
  }

  clear(): void {
    this.entries = [];
  }

  clone(): WeightedRandomBag<T> {
    const bag = new WeightedRandomBag<T>();
    for (const e of this.entries) {
      bag.add(e.weight, e.item);
    }
    return bag;
  }

  merge(other: WeightedRandomBag<T>): this {
    for (const otherEntry of other.entries) {
      const existing = this.entries.find(e => e.item === otherEntry.item);
      if (existing) {
        existing.weight += otherEntry.weight;
      } else {
        this.entries.push({ ...otherEntry });
      }
    }
    return this;
  }

  get size(): number {
    return this.entries.length;
  }
}
